### PR TITLE
[runtime] Don't insta-fail when a faulty COM type is encountered.

### PR DESCRIPTION
**Pull request by vkargov**
This patch is meant to fix bug 8477.
Currently if Mono encounters some managed code trying to interact with an invalid COM type, it will consider it a fatal error (g_assert).
However, that bad code may be hidden in a basic block that is never really executed.
This patch is meant to replace the compile-time assertion with a runtime safeguard for all invalid COM invocations it encounters. (That safeguard should never be reached either since if a COM object is invalid, an exception should be thrown during its creation.)

### CODE REVIEW

File: mono/metadata/cominterop.c [hunk 1]
```diff
@@ -378,21 +378,18 @@ cominterop_get_method_interface (MonoMethod* method)
...
  * @method: a method
+ * @error: pointer to a bool indicating whether the lookup was successful or not; all failures will be considered fatal if it's set to null.
  *
  * Returns: the method's slot in the COM interface vtable
  */
 static int
-cominterop_get_com_slot_for_method (MonoMethod* method)
+cominterop_get_com_slot_for_method (MonoMethod* method, gboolean* error)
 {
 	guint32 slot = method->slot;
```
Comment (vargaz) [inline mono/metadata/cominterop.c]: This should use a MonoError.

**Comment by akoeplinger** [comment]
@monojenkins build Linux x64 mcs

File: mono/metadata/cominterop.c [hunk 2]
```diff
@@ -643,11 +648,24 @@ void
...
+		mono_mb_emit_icall (mb, cominterop_get_function_pointer);
+		/* Leaves the function pointer on top of the stack */
+	}
+	else {
+		/*
+		 * This code should never be executed and all invalid COM objects should be sieved out during their creation.
+		 * However, it will be emitted as a safeguard for cases when a method tries to interact with an invalid COM object from an unreachable basic block.
+		 */
+		char *msg = g_strdup_printf ("Unable to locate the interface on which method %s is defined.", mono_method_get_full_name (method));
+		mono_mb_emit_exception (mb, "SystemException", msg);
```
Comment (kumpera) [inline mono/metadata/cominterop.c]: This looks like the wrong exception, how did you come to use it?

File: mono/tests/bug-8477.il [hunk 3]
```diff
@@ -0,0 +1,115 @@
+// This test is meant to make sure Mono doesn't fail when invalid COM invocations are present but never executed.
+// See https://bugzilla.xamarin.com/show_bug.cgi?id=8477 for details.
```
Comment (kumpera) [inline mono/tests/bug-8477.il]: Why does this needs to be an IL test?

Comment (vkargov) [reply mono/tests/bug-8477.il]: The original bug report provides two assemblies, an exe and a dll. I figured the most transparent way of making a test would be converting it into the IL form, cleaning it up a little and merging together into a single IL file. What test format would you suggest that I use?

File: mono/metadata/cominterop.c (hunk 2 above)
Comment (vkargov) [reply mono/metadata/cominterop.c]: You're right, `System.Runtime.InteropServices.InvalidComObjectException` should suit it better, because invalid COM objects are what the code is meant to handle.
There's one caveat though which makes me a bit uncertain. Namely, this branch is never really meant to be executed since all invalid COM objects be spotted during their creation, so in this regard I think it would make better sense to throw `System.ExecutionEngineException`.
I'm curious what you think about it.

File: mono/tests/bug-8477.il (hunk 3 above)
Comment (kumpera) [reply mono/tests/bug-8477.il]: C#

File: mono/metadata/cominterop.c (hunk 1 above)
Comment (vkargov) [reply mono/metadata/cominterop.c]: I've switched to MonoError, please let me know if there's something wrong with my implementation.

File: mono/tests/bug-8477.il (hunk 3 above)
Comment (vkargov) [reply mono/tests/bug-8477.il]: Done, though to be honest, I think rewriting an IL test case in C# has a chance of defeating its initial purpose. Newer versions of csc/mcs may change the way they handle some edge cases in the future , thus producing assemblies that won't check for what they were meant to check in the first place. For instance, they could "optimize too much" and eliminate a branch that is never taken but needs to be there to trigger the bug, or detect an invalid COM type during compilation.

File: mono/metadata/cominterop.c [hunk 4]
```diff
@@ -378,30 +378,33 @@ cominterop_get_method_interface (MonoMethod* method)
...
+	error_init (error);
+
 	/* if method is on a class, we need to look up interface method exists on */
 	if (!MONO_CLASS_IS_INTERFACE(ic)) {
 		int offset = 0;
 		int i = 0;
 		ic = cominterop_get_method_interface (method);
+		if (!ic || !MONO_CLASS_IS_INTERFACE (ic)) {
+			mono_error_set_invalid_program (error, "COM interface for method %s cannot be found.", mono_method_full_name (method, TRUE));
+			return -1;
```
Comment (vargaz) [inline mono/metadata/cominterop.c]: This will leak the result of mono_method_full_name (), you should do:
char *name = mono_method_full_name ();
set_invalid_program (...)
g_free (name);

Comment (vkargov) [reply mono/metadata/cominterop.c]: Good catch! Fixed and updated.

File: mono/metadata/cominterop.c [hunk 5]
```diff
@@ -643,11 +648,25 @@ void
...
+		mono_mb_emit_icon (mb, slot);
+		mono_mb_emit_icall (mb, cominterop_get_function_pointer);
+		/* Leaves the function pointer on top of the stack */
+	}
+	else {
+		/*
+		 * This code should never be executed and all invalid COM objects should be sieved out during their creation.
+		 * However, it will be emitted as a safeguard for cases when a method tries to interact with an invalid COM object from an unreachable basic block.
+		 */
+		char *msg = g_strdup_printf ("Unable to locate the interface on which method %s is defined.", mono_method_get_full_name (method));
```
Comment (luhenry) [inline mono/metadata/cominterop.c]: You are completely missing the point of MonoError: you are setting the error in `cominterop_get_com_slot_for_method` only not to use it here. Please use the exception from the MonoError instead of creating a new one from scratch.

Comment (vkargov) [reply mono/metadata/cominterop.c]: I'm not sure that is the case here.
The error created inside `cominterop_get_com_slot_for_method` signifies that the COM type is invalid, which means there's something wrong with the program being executed by Mono. That is "normal" in the sense that even though the program has an invalid COM type, the runtime has spotted it timely and predictably. Thus, an error generated within `cominterop_get_com_slot_for_method` indicates there's something wrong with the program being executed, but not the runtime. 
This line, however, generates a safeguard that effectively marks this basic block as "bad" because we've deduced that a method on an invalid COM object is invoked from it. An exception should've been thrown during the creation of an invalid COM object, so this block should not have been reached. If we do find ourselves here, that means something went wrong in the runtime, which sounds like a case of `ExecutionEngineException` to me. I use this exception as an assertion here.
Perhaps something is wrong with my reasoning, in which case I'll be grateful if you point out what exactly is wrong. Admittedly I've not used that mechanism much. 
The way I see it right now, it would be logically wrong to use the same kind of `MonoError` in those two functions, and that is the reason why my initial patch had a simple boolean error code instead of `MonoError` in the first place.

Comment (luhenry) [reply mono/metadata/cominterop.c]: You can just use the `InvalidProgramException`.

Comment (vkargov) [reply mono/metadata/cominterop.c]: I found out that CLR throws `InvalidOperationException`, so I'll update the patch to emit that one instead.
Furthermore, I found a case where this stub will be reached under "normal" circumstances (namely, if the incomplete method of a ComImport class is static) rendering my original presumption ("this should never be reached") incorrect. Thus, the exception does need to be the same and my previous.
I'll update the PR soon with these changes.

**Comment by vkargov** [comment]
I found a similar case involving failures on "meaningless" static `ComImport` types, so I expanded the test and the patch to handle it as well. I also added a couple new functions since I couldn't find a existing way to emit an exception with `MonoBuilder` for a `MonoError`.
Let me know if I missed something.
